src/FindEmailAddress.java

```diff
@@ -0,0 +1,362 @@
+/*
+*
+* The project is for finding email address in the web site
+*
+* Create a command line program that will take an internet domain name (i.e. “jana.com”) and
+* print out a list of the email addresses that were found on that website only.
+* the application will find email addresses on any discoverable page of the website.
+*
+* We implemented a multithread method. each url request and email&href check is a threadTask
+*
+*
+*/
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Hashtable;
+import java.net.URI;
+import java.util.concurrent.*;
+
+/*
+*  This is the main class to get the email addresses
+*
+*/
+public class FindEmailAddress {
+
+    private String rootURL;
+    public String requestDomain;
+    public Hashtable<String,String> urlHistory;
+    public Hashtable<String,String> emails;
+    public static int maxUrlCheck = 200; //to avoid too many pages to check we add maximum number of url to check, it can be modified
+    public ExecutorService executor;
+    public Integer tasksNumber;
+
+    public static void main(String[] args)
+    {
+        String requestUrl = null;
+
+        if(args.length >0)
+        {
+            requestUrl =  args[0];
+        }
+        if (requestUrl == null)
+            return;
+
+        FindEmailAddress findEmailAddress = new FindEmailAddress(requestUrl);
+        findEmailAddress.startWebCrawl();
+        findEmailAddress.printout();
+    }
+
+    /*
+    *  @param requestURL is the root url. All the webpage with same domain will be searched for email address
+    */
+    public FindEmailAddress(String requestURL)
+    {
+        this.rootURL = requestURL;
+        if(!requestURL.toLowerCase().startsWith("http"))
+        {
+            rootURL = "http://" + requestURL;
+        }
+        try {
+            URI uri = new URI(rootURL);
+            this.requestDomain= uri.getHost();
+
+        }catch(java.net.URISyntaxException e)
+        {
+            e.printStackTrace();
+        }
+        this.emails = new Hashtable<String,String>();
+        this.urlHistory = new Hashtable<String, String>();
+        this.urlHistory.put(rootURL,"");
+        tasksNumber = 0;
+    }
+
+    /*
+     *  This is the function to do web crawling
+     *  it will create a thread pool
+     *  if all the tasks are done, it will stop.
+     */
+
+    public void startWebCrawl()
+    {
+        if(rootURL == null || rootURL.trim().length()<=0)
+            return;
+
+        executor = Executors.newFixedThreadPool(10);
+        executor.execute(new Handler(rootURL));
+        do{
+            try {
+                Thread.sleep(3000);
+            }catch(Exception e)
+            {
+                e.printStackTrace();
+            }
+        }while(this.tasksNumber >0);
```

Comment:
I can see what you're doing here, and it's smart to try to get some parallelization in, but the `sleep(3000)` is strange; why 3000 and not some other number? do you even need this thread at all? it's not doing anything. can't the `Handler` threads know when they are done?

Reply:
I think there's a bug in this logic as well; `tasksNumber` should start at 1, not 0.

Reply:
This is a good question. I took really some time trying to find better solution, I couldn't. I can re-think about it.
3 second sleep time is just to avoid main thread taking too much CPU time, it is just waiting for all the tasks to be done. It can be any other reasonable number. The drawback of this will delay the final result.
We created 10 threads in thread pool (ExecutorService), these threads will be terminated until we call executor.shutdown(). These loop is just to block the main thread until the tasksNumber to be 0. Then the next line code shutdown() will be executed.

Reply:
Yes, see the issue, should start with 1, thanks.

Reply:
Ok, I find a better solution, I am using Object.wait() to block the main thread, and when taskNumber == 0, it will call Object.notify() to release the block. This solution is using notification method much better than loop checking. I checked in the code. and thanks again.